### PR TITLE
IGNITE-17139 Fix creation of new RocksDB data regions and add examples

### PR DESCRIPTION
**Pull request by sashapolo**
https://issues.apache.org/jira/browse/IGNITE-17139

### CODE REVIEW

**Review by rpuch — APPROVED** [review]

File: modules/storage-rocksdb/src/test/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                            RocksDbDataStorageConfigurationSchema.class
+                    }
+            ) TableConfiguration tableCfg
+    ) {
+        TableStorage table = engine.createTable(tableCfg);
+
+        table.start();
+
+        try {
+            var dataStorageConfig = (RocksDbDataStorageConfiguration) table.configuration().dataStorage();
```
Comment (SammyVimes) [inline modules/storage-rocksdb/src/test/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineTest.java:81]: var!

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: examples/src/main/java/org/apache/ignite/example/storage/RocksDbStorageExample.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ *     <li>
+ *         Start an Ignite node using the CLI tool:<br>
+ *         {@code ignite node start --config=$IGNITE_HOME/examples/config/ignite-config.json my-first-node}
+ *     </li>
+ *     <li>
+ *         Cluster initialization using the CLI tool (if not done yet):<br>
+ *         {@code ignite cluster init --cluster-name=ignite-cluster --node-endpoint=localhost:10300 --meta-storage-node=my-first-node}
+ *     </li>
+ *     <li>
+ *         Add configuration for in-memory data region of of the PageMemory storage engine using the CLI tool (if not done yet):<br>
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply examples/src/main/java/org/apache/ignite/example/storage/RocksDbStorageExample.java]: found that myself too)

File: modules/storage-rocksdb/src/test/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                .change(c -> c.create(customRegionName, rocksDbDataRegionChange -> {}));
+
+        assertThat(engineConfigChangeFuture, willCompleteSuccessfully());
+
+        TableStorage table = engine.createTable(tableCfg);
+
+        table.start();
+
+        try {
+            var dataStorageConfig = (RocksDbDataStorageConfiguration) table.configuration().dataStorage();
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/storage-rocksdb/src/test/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineTest.java:114]: what's wrong with that?

File: modules/storage-rocksdb/src/test/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineTest.java (hunk 1 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/test/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineTest.java:81]: what's wrong with that?

**Review by SammyVimes — APPROVED** [review]
LGTM!

**Comment by rpuch** [comment]
Looks good to me

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java [hunk 4]
```diff
@@ -75,19 +78,27 @@ public RocksDbStorageEngine(RocksDbStorageEngineConfiguration engineConfig, Path
...
+                return CompletableFuture.completedFuture(null);
+            }
+        });
+    }
 
-        regions.put(DEFAULT_DATA_REGION_NAME, defaultRegion);
+    private void registerDataRegion(RocksDbDataRegionConfiguration dataRegionConfig) {
+        var region = new RocksDbDataRegion(dataRegionConfig);
 
-        for (String regionName : engineConfig.regions().value().namedListKeys()) {
```
Comment (SammyVimes) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java:84]: This shouldn't be removed, we still need to create preconfigured regions

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java (hunk 4 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java:84]: Actually, no. When creating an Ignite instance, listeners get notified on start, please see the `IgniteImpl#notifyConfigurationListeners` method

**Comment by SammyVimes** [comment]
Thank you for the contribution, merged:

`main` branch – d25a08941a3a415d5586af5a456d3ffaf25afa8d
`ignite-3.0.0-alpha5` branch – 61649d2942bbeb99ef29e615b42d83d548d9b242

**Pull request closed without merging**